[fix]: strip leading newline after `<pre>` and `<textarea>`

ota-meshi opened the pull request:
### Before submitting the PR, please make sure you do the following
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] Prefix your PR title with `[feat]`, `[fix]`, `[chore]`, or `[docs]`.
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
-  [x] Run the tests with `npm test` and lint the project with `npm run lint`


Close #7264


This PR changes to strip the newline immediately after `<pre>`.
However, if `innerHTML` is used in the renderer and there are multiple newlines, the newline immediately after `<pre>` is restored.

~~However, the SSR remains the same as before. This isn't necessary because HTML parsing strips the newline immediately after `<pre>`, and it's simpler than handling multiple newlines.~~
It has been changed to handle the same in SSR.

ota-meshi commented:
I don't know how to fix the CI timeout. If you think my change is the cause, please let me know.

tanhauhau commented on src/compiler/compile/render_dom/wrappers/Element/index.ts:
```diff
@@ -1015,6 +1015,15 @@ function to_html(wrappers: Array<ElementWrapper | TextWrapper | MustacheTagWrapp
 
+				if (wrapper.node.name === 'pre') {
+					// Two or more leading newlines are required to restore the leading newline immediately after `<pre>`.
+					// see https://html.spec.whatwg.org/multipage/grouping-content.html#the-pre-element
```
from https://html.spec.whatwg.org/multipage/syntax.html#element-restrictions

i think we should add the same logic for `<textarea>` as well

ota-meshi replied:
I didn't know that. Thanks for letting me know.

tanhauhau commented on test/runtime/samples/pre-tag/.editorconfig:
```diff
@@ -0,0 +1,2 @@
+[main.svelte]
+trim_trailing_whitespace = unset
```
?
remove the extra file?

ota-meshi replied:
This is necessary to edit the blank test code before the line break in the editor.

https://github.com/sveltejs/svelte/pull/7280/files#diff-d5881bd31993ce2fe17bf9174468471046fe3c66fb8c1cf2b43a842aa41fbd77R49-R50

Without it, the test code will change unintentionally when someone edits the file in the future.

tanhauhau replied:
i wonder is it there's some plugins that do some auto formatting before u save?

in any case, the unintentional changes will lead to git changes / test failed right?

ota-meshi replied:
You are using EditorConfig. There is a configuration file at the root of the project.

https://github.com/sveltejs/svelte/blob/master/.editorconfig

The added file is for overwriting its configuration.

ota-meshi replied:
> in any case, the unintentional changes will lead to git changes / test failed right?

If trailing whitespace is removed from the .svelte file, the test on the next line will fail.

https://github.com/sveltejs/svelte/pull/7280/files#diff-0a1ca6ee0ecfd38066746f565bce75c4f5ce64d2b5ac5505330c2e4a85769547R52

tanhauhau commented on test/server-side-rendering/samples/pre-tag/.editorconfig:
```diff
@@ -0,0 +1,2 @@
+[{main.svelte,_expected.html}]
+trim_trailing_whitespace = unset
```
extra file?

tanhauhau commented on src/compiler/compile/nodes/Text.ts:
```diff
@@ -49,6 +49,15 @@ export default class Text extends Node {
+	should_strip_leading_newline(): boolean {
+		const parent = this.parent;
+		return parent.type === 'Element' && parent.name === 'pre' && parent.children[0] === this;
+	}
```
similar to https://github.com/sveltejs/svelte/pull/7280/files#r845152518

ota-meshi commented:
@tanhauhau Thank you for checking this PR!

I made a fix for `<textarea>`.
However, since `<textarea>` has content in the `value` attribute, I added a different logic than `<pre>`.
Also, I changed it to use the content instead of the `value` attribute when using `innerHTML`.
Also, SSR will always include the newline after `<textarea>`. It is possible to reduce newlines by checking for the presence of the first newline, but I thought that the logic would be complicated, so I changed it to always put a newline.

ota-meshi commented:
> Also, I changed it to use the content instead of the value attribute when using innerHTML.

Currently, the following JS is generated.

```svelte
<div>
	<textarea>AAA</textarea>
</div>
```

```js
div.innerHTML = `<textarea value="AAA"></textarea>`;
```

I don't think this should work, but it works fine with REPL. (I didn't understand why it worked.)

https://svelte.dev/repl/d5ccce1d50d547039053b78db9ce8477?version=3.46.6

However, it didn't work in the test, so I changed it to render to the content.

```js
div.innerHTML = `<textarea>AAA</textarea>`;
```

tanhauhau commented on src/compiler/compile/render_ssr/handlers/Element.ts:
```diff
@@ -164,6 +164,11 @@ export default function(node: Element, renderer: Renderer, options: RenderOption
+			if (node.name === 'textarea') {
+				// Restore the leading newline immediately after `<textarea>`.
+				// see https://html.spec.whatwg.org/multipage/syntax.html#element-restrictions
+				renderer.add_string('\n');
```
this seemed to indiscriminately add newline even when the original code did not have a newline character

ota-meshi replied:
Yes. That's right. The result displayed in the browser does not change, so there should be no problem other than the file size.
I thought that complicated source code was needed to reduce unnecessary line feed codes. Do you know a good way?

tanhauhau replied:
what i meant is that in line 180, for `<pre>`, you only prepended `'\n'` if the `first.data` starts with newline, i believe `<textarea>` need a similar condition right?

in the `server-side-rendering/samples/textarea-content` test case, the `#textarea-without-leading-newline` textarea didnt have newline in the source but ended up having newlines in the rendered HTML instead

ota-meshi replied:
Oh thank you. I understand what you say. I thought I had to use node_contents to determine the starts with newline, but as you say it just accesses the original attribute node.

tanhauhau commented:
> I don't think this should work, but it works fine with REPL. (I didn't understand why it worked.)

oh the result u see in REPL uses `dev` mode to build, in which it does not have the `.innerHTML` optimisation.

ota-meshi commented:
> oh the result u see in REPL uses dev mode to build, in which it does not have the .innerHTML optimisation.

Ah I see. Thank you for letting me know.
Should I open the PR as another bug?
(But, I don't think anyone actually uses fixed strings in `<textarea>`.)

tanhauhau commented:
> Should I open the PR as another bug?

i see you already made relevant changes right, it's okay i'll review them later

tanhauhau commented on src/compiler/compile/render_dom/wrappers/Element/index.ts:
```diff
@@ -1015,6 +1021,41 @@ function to_html(wrappers: Array<ElementWrapper | TextWrapper | MustacheTagWrapp
+								};
+							}
+						});
+					}
```
should be able to reuse `to_html` ?

something like `to_html(value_attribute.node.chunks, block, literal, state)` ?

ota-meshi replied:
I don't think we can use `to_html`, because `value_attribute.node.chunks` is an array of not wrapper nodes.

But since it's the same as L1004, do you think I should add a function and use it?
https://github.com/sveltejs/svelte/pull/7280/files#diff-d6a6e8b36b1b88c9bd08d5bee1887bfffc7cebb2211ce45591ef59f8d2fc1111R1004

tanhauhau replied:
oh yes 🙂

ota-meshi commented on src/compiler/compile/render_dom/wrappers/Fragment.ts:
```diff
@@ -127,8 +128,12 @@ export default class FragmentWrapper {
+					first.data = trim_start(first.data);
+				} else if (first.node.should_strip_leading_newline()) {
+					first.data = first.data.replace(start_newline, '');
+				}
```
Thank you for checking.
I will make that change.

tanhauhau replied:
is it okay to revert the changes on this file to make the diff cleaner?

since you no longer add a new condition here

ota-meshi commented:
Thank you for checking this PR.
I changed this PR. So please check again.

The CI error seems to be due to a timeout. If you think my change is the cause, please let me know.
